### PR TITLE
Tests have main linter

### PR DESCRIPTION
**Pull request by clee2000**
The linter uses libcst to check for a call to run_tests or a raised exception when the test file is run as main to ensure that all test files either get run in OSS CI or don't run and are expected to not run.

A better option instead of making this into a linter might be to add this code in run_test since there's also a list of blocklisted tests there that needs to be updated when a test file raises an exception.

This is possibly overkill since run on its own, the code takes ~1 minutes to run without the multiprocessing on all the files

### CODE REVIEW

**Review by kit1980 — APPROVED** [review]

File: tools/linter/adapters/test_has_main_linter.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+            name,
+            [m.ComparisonTarget(m.Equal(), main)],
+        )
+        if_main2 = m.Comparison(
+            main,
+            [m.ComparisonTarget(m.Equal(), name)],
+        )
+        for child in node.children:
+            if m.matches(child, m.If(test=if_main1 | if_main2)):
+                if len(m.findall(node, run_test_call | raise_block)) > 0:
```
Comment (kit1980) [inline tools/linter/adapters/test_has_main_linter.py:45]: This should also work, right?

```suggestion
                if m.findall(node, run_test_call | raise_block):
```

File: tools/linter/adapters/test_has_main_linter.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+
+LINTER_CODE = "TEST_HAS_MAIN"
+
+
+class HasMainVisiter(cst.CSTVisitor):
+    def __init__(self) -> None:
+        super().__init__()
+        self.found = False
+
+    def visit_Module(self, node: cst.Module) -> bool:
```
Comment (kit1980) [inline tools/linter/adapters/test_has_main_linter.py:26]: Rename `node` argument? As `cst.Module` is not a `cst.Node`.

File: tools/linter/adapters/test_has_main_linter.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+        # name == main or main == name
+        if_main1 = m.Comparison(
+            name,
+            [m.ComparisonTarget(m.Equal(), main)],
+        )
+        if_main2 = m.Comparison(
+            main,
+            [m.ComparisonTarget(m.Equal(), name)],
+        )
```
[the remark this answers is not in the document]
Comment (clee2000) [reply tools/linter/adapters/test_has_main_linter.py:42]: i tried it out but there doesn't seem to be a significant difference unfortunately

**Pull request closed without merging**